aydin/nn/models/torch/torch_jinet.py

```diff
@@ -0,0 +1,280 @@
+from collections import OrderedDict
+from itertools import chain
+
+import torch
+from torch import nn
+from torch.optim.lr_scheduler import ReduceLROnPlateau
+from torch.utils.tensorboard import SummaryWriter
+
+from aydin.nn.layers.dilated_conv import DilatedConv
+from aydin.nn.pytorch.optimizers.esadam import ESAdam
+from aydin.util.log.log import lprint
+
+
+class JINetModel(nn.Module):
+    def __init__(
+        self,
+        spacetime_ndim,
+        nb_out_channels: int = 1,
+        kernel_sizes=None,
+        num_features=None,
+        nb_dense_layers: int = 3,
+        nb_channels: int = None,
+        final_relu: bool = False,
+    ):
+        super(JINetModel, self).__init__()
+
+        self.spacetime_ndim = spacetime_ndim
+        self.nb_out_channels = nb_out_channels
+        self._kernel_sizes = kernel_sizes
+        self._num_features = num_features
+        self.nb_dense_layers = nb_dense_layers
+        self.nb_channels = nb_channels
+        self.final_relu = final_relu
+
+        if len(self.kernel_sizes) != len(self.num_features):
+            raise ValueError("Number of kernel sizes and features does not match.")
+
+        self.dilated_conv_functions = []
+        current_receptive_field_radius = 0
+        for scale_index in range(len(self.kernel_sizes)):
+            # Get kernel size and number of features:
+            kernel_size = self.kernel_sizes[scale_index]
+
+            # radius and dilation:
+            radius = (kernel_size - 1) // 2
+            dilation = 1 + current_receptive_field_radius
+
+            self.dilated_conv_functions.append(
+                DilatedConv(
+                    1 if scale_index == 0 else self.num_features[scale_index - 1],
+                    self.num_features[scale_index],
+                    self.spacetime_ndim,
+                    padding=dilation * radius,
+                    kernel_size=kernel_size,
+                    dilation=dilation,
+                    activation="lrel",
+                )
+            )
+
+            # update receptive field radius
+            current_receptive_field_radius += dilation * radius
+
+        if spacetime_ndim == 2:
+            self.conv = nn.Conv2d
+        elif spacetime_ndim == 3:
+            self.conv = nn.Conv3d
+        else:
+            raise ValueError("spacetime_ndim can not be anything other than 2 or 3...")
+
+        if self.nb_channels is None:
+            self.nb_channels = sum(self.num_features)  # * 2
```

Comment:
what's the logic behind the multiplying by 2 here @li-li-github ?

Reply:
I think that's a heritage from Loic's original code and I don't know why. And I guess I comment if out b/c it takes up too much memory but no significant difference in output.